tests/topotests/bgp_vpnv4_noretain/test_bgp_vpnv4_noretain.py

```diff
@@ -133,27 +133,37 @@ def teardown_module(_mod):
     tgen.stop_topology()
 
 
-def test_protocols_convergence():
+def router_json_cmp_exact_filter(router, cmd, expected):
+    # filter out tableVersion, version and nhVrfID
+    output = router.cmd('vtysh -c "{}" | grep -v ersion | grep -v nhVrfId'.format(cmd))
```

Comment:
Why is this needed? Can't we just drop those fields from the `json_file`?

Reply:
Already dropped from the `json_file` but this is required for exact match

Reply:
I'd suggest using then not with the files, but with the Python dicts. Like we have already in lots of places:

```python
    def _bgp_aggregated_summary_only_med_match():
        output = json.loads(r3.vtysh_cmd("show bgp ipv4 unicast json"))
        expected = {
            "routes": {
                "172.16.255.0/24": [{"path": "65002"}],
                "172.16.255.1/32": None,
                "172.16.255.2/32": None,
                "172.16.255.3/32": None,
            }
        }
        return topotest.json_cmp(output, expected)
```
If you expect only 172.16.255.0/24, then others are checked against None then. 

Can't this be adopted like in this example or have I missed something more?

Reply:
It does not work with exact match:
```
2023-01-24 10:43:30,741 ERROR: topolog: 'router_json_cmp' failed after 5.47 seconds
2023-01-24 10:43:30,742 ERROR: topolog: test failed at "bgp_vpnv4_noretain.test_bgp_vpnv4_noretain/test_bgp_no_retain": "r1" JSON output mismatches
assert Generated JSON diff error report:
  
  > $->tableVersion: d1 has element with value '2' but in d2 it has value 'None'
  > $->routes->routeDistinguishers->444:1->10.201.0.0/24[0]->version: d1 has element with value '2' but in d2 it has value 'None'
  > $->routes->routeDistinguishers->444:1->10.201.0.0/24[0]->nhVrfId: d1 has element with value '7' but in d2 it has value 'None'
  > $->routes->routeDistinguishers->444:2->10.200.0.0/24[0]->version: d1 has element with value '1' but in d2 it has value 'None'
```

Here is the patch:

```
diff --git a/tests/topotests/bgp_vpnv4_noretain/r1/ipv4_vpn_routes.json b/tests/topotests/bgp_vpnv4_noretain/r1/ipv4_vpn_routes.json
index 28e153e3de..d822965e2b 100644
--- a/tests/topotests/bgp_vpnv4_noretain/r1/ipv4_vpn_routes.json
+++ b/tests/topotests/bgp_vpnv4_noretain/r1/ipv4_vpn_routes.json
@@ -1,6 +1,7 @@
 {
    "vrfId":0,
    "vrfName":"default",
+   "tableVersion":null,
    "routerId":"1.1.1.1",
    "defaultLocPrf":100,
    "localAS":65500,
@@ -16,6 +17,7 @@
                   "prefix":"10.201.0.0",
                   "prefixLen":24,
                   "network":"10.201.0.0\/24",
+                  "version":null,
                   "metric":0,
                   "weight":32768,
                   "peerId":"(unspec)",
@@ -23,6 +25,7 @@
                   "origin":"incomplete",
                   "announceNexthopSelf":true,
                   "nhVrfName":"vrf1",
+                  "nhVrfId":null,
                   "nexthops":[
                      {
                         "ip":"0.0.0.0",
@@ -44,6 +47,7 @@
                   "prefix":"10.200.0.0",
                   "prefixLen":24,
                   "network":"10.200.0.0\/24",
+                  "version":null,
                   "metric":0,
                   "locPrf":100,
                   "weight":0,
diff --git a/tests/topotests/bgp_vpnv4_noretain/r1/ipv4_vpn_routes_unfiltered.json b/tests/topotests/bgp_vpnv4_noretain/r1/ipv4_vpn_routes_unfiltered.json
index 45f4acce6f..b27007810e 100644
--- a/tests/topotests/bgp_vpnv4_noretain/r1/ipv4_vpn_routes_unfiltered.json
+++ b/tests/topotests/bgp_vpnv4_noretain/r1/ipv4_vpn_routes_unfiltered.json
@@ -1,6 +1,7 @@
 {
    "vrfId":0,
    "vrfName":"default",
+   "tableVersion":null,
    "routerId":"1.1.1.1",
    "defaultLocPrf":100,
    "localAS":65500,
@@ -16,6 +17,7 @@
                   "prefix":"10.201.0.0",
                   "prefixLen":24,
                   "network":"10.201.0.0\/24",
+                  "version":null,
                   "metric":0,
                   "weight":32768,
                   "peerId":"(unspec)",
@@ -23,6 +25,7 @@
                   "origin":"incomplete",
                   "announceNexthopSelf":true,
                   "nhVrfName":"vrf1",
+                  "nhVrfId":null,
                   "nexthops":[
                      {
                         "ip":"0.0.0.0",
@@ -44,6 +47,7 @@
                   "prefix":"10.200.0.0",
                   "prefixLen":24,
                   "network":"10.200.0.0\/24",
+                  "version":null,
                   "metric":0,
                   "locPrf":100,
                   "weight":0,
@@ -71,6 +75,7 @@
                   "prefix":"10.210.0.0",
                   "prefixLen":24,
                   "network":"10.210.0.0\/24",
+                  "version":null,
                   "metric":0,
                   "locPrf":100,
                   "weight":0,
diff --git a/tests/topotests/bgp_vpnv4_noretain/test_bgp_vpnv4_noretain.py b/tests/topotests/bgp_vpnv4_noretain/test_bgp_vpnv4_noretain.py
index ed5cc3faf6..33de4c9b3d 100644
--- a/tests/topotests/bgp_vpnv4_noretain/test_bgp_vpnv4_noretain.py
+++ b/tests/topotests/bgp_vpnv4_noretain/test_bgp_vpnv4_noretain.py
@@ -134,8 +134,7 @@ def teardown_module(_mod):
 
 
 def router_json_cmp_exact_filter(router, cmd, expected):
-    # filter out tableVersion, version and nhVrfID
-    output = router.cmd('vtysh -c "{}" | grep -v ersion | grep -v nhVrfId'.format(cmd))
+    output = router.cmd('vtysh -c "{}"'.format(cmd))
     logger.info("{}: {}\n{}".format(router.name, cmd, output))
 
     json_output = json.loads(output)
@@ -163,10 +162,11 @@ def test_bgp_no_retain():
 
     expected = json.loads(open(json_file).read())
     test_func = partial(
-        router_json_cmp_exact_filter,
+        topotest.router_json_cmp,
         router,
         "show bgp ipv4 vpn json",
         expected,
+        exact=True,
     )
     _, result = topotest.run_and_expect(test_func, None, count=10, wait=0.5)
     assertmsg = '"{}" JSON output mismatches'.format(router.name)
@@ -196,10 +196,11 @@ def test_bgp_retain():
 
     expected = json.loads(open(json_file).read())
     test_func = partial(
-        router_json_cmp_exact_filter,
+        topotest.router_json_cmp,
         router,
         "show bgp ipv4 vpn json",
         expected,
+        exact=True,
     )
     _, result = topotest.run_and_expect(test_func, None, count=10, wait=0.5)
     assertmsg = '"{}" JSON output mismatches'.format(router.name)

```

Reply:
Yes, but what happens if you don't use JSON, but Python dict?

Reply:
What do you mean ? JSON is converted to Python dict ?